File: lib/src/verifier.ts
/**
 * v0.1: top-level verifier entry.
 *
 * Pipeline (mirrors the Java reference):
 *   1. Unzip the .aep package.
 *   2. Read required entries (response.txt, canonical.bin, hash.sha256,
 *      signature.sig, public_key.pem, metadata.json, timestamp.tsr).
 *   3. Recompute supported canonical forms and compare to canonical.bin.
 *   4. Hash canonical bytes with SHA-256; compare to hash.sha256.
 *   5. Verify RSA signature with public_key.pem.
 *   6. If PQC entries present, verify ML-DSA-65 signature.
 *   7. Structural-check the RFC 3161 timestamp.
 *
 * Each step appends to the report; a single failure short-circuits.
 */

import { unzipSync } from "fflate";

import { canonical, jcs } from "./canonical.js";
import { sha256, toHex } from "./hash.js";
import { decodeBase64, importRsaPublicKey, verifyRsa, verifyRsaDigestInfo } from "./rsa.js";
import { verifyMlDsa65 } from "./mldsa.js";
import { inspectTsa, verifyTsaTrust } from "./tsa.js";
import { parseAndValidateOvertReceipt, type OvertReceipt } from "./overt.js";
import type { FailureCode, VerifyOptions, VerifyResult } from "./index.js";

const TEXT_DEC = new TextDecoder();
const MAX_ARCHIVE_BYTES = 64 * 1024 * 1024;
const MAX_ENTRY_BYTES = 16 * 1024 * 1024;
const MAX_EXPANDED_BYTES = 32 * 1024 * 1024;
const MAX_ENTRIES = 32;

export async function verify(
  input: Uint8Array | ArrayBuffer | Blob,
  opts: VerifyOptions = {},
): Promise<VerifyResult> {
  const report: string[] = [];
  const bytes = await toBytes(input);
  let metadata: Record<string, unknown> | null = null;
  let overtReceipt: OvertReceipt | null = null;

  if (bytes.byteLength > MAX_ARCHIVE_BYTES) {
    return fail(
      report,
      "ZIP_ARCHIVE_LIMIT",
      "Package exceeds the 64 MiB archive safety limit.",
      metadata,
    );
  }

  let entries: Record<string, Uint8Array>;
  try {
    let entryCount = 0;
    let expandedBytes = 0;
    const names = new Set<string>();
    entries = unzipSync(bytes, {
      filter: (file) => {
        entryCount += 1;
        expandedBytes += file.originalSize;
        if (entryCount > MAX_ENTRIES) throw new Error("too many ZIP entries");
        if (file.originalSize > MAX_ENTRY_BYTES) throw new Error("ZIP entry too large");
        if (expandedBytes > MAX_EXPANDED_BYTES) throw new Error("expanded ZIP too large");
        if (
          file.name.includes("/") ||
          file.name.includes("\\") ||
          file.name.includes("\0")
        ) {
          throw new Error("AEP entries must use flat, safe names");
        }
        if (names.has(file.name)) throw new Error("duplicate ZIP entry");
        names.add(file.name);
        return true;
      },
    });
  } catch (e) {
    return fail(
      report,
      "ZIP_INVALID_OR_UNSAFE",
      `Package failed ZIP parsing or safety limits: ${(e as Error).message}.`,
      metadata,
    );
  }
  report.push(`Package unzipped (${Object.keys(entries).length} entries).`);

  // Required entries.
  const required = ["response.txt", "canonical.bin", "hash.sha256", "signature.sig", "public_key.pem", "metadata.json", "timestamp.tsr"];
  for (const name of required) {
    if (!entries[name]) {
      return fail(
        report,
        "REQUIRED_ENTRY_MISSING",
        `Missing required entry: ${name}.`,
        metadata,
      );
    }
  }

  // Parse metadata.json for reporting + re-canonicalisation.
  try {
    const parsed = JSON.parse(TEXT_DEC.decode(entries["metadata.json"]!)) as unknown;
    if (parsed === null || typeof parsed !== "object" || Array.isArray(parsed)) {
      return fail(
        report,
        "METADATA_NOT_OBJECT",
        "metadata.json is not a JSON object.",
        metadata,
      );
    }
    metadata = parsed as Record<string, unknown>;
  } catch {
    return fail(
      report,
      "METADATA_INVALID_JSON",
      "metadata.json is not valid JSON.",
      metadata,
    );
  }
  report.push("metadata.json parsed.");

  // Recompute supported canonical forms. Current signers use response + LF +
  // JCS(metadata). Response-only verification is read-only compatibility for
  // packages produced by early Java tooling; it does not authenticate metadata.
  let profileCanonical: Uint8Array;
  try {
    profileCanonical = canonical({
      responseBytes: entries["response.txt"]!,
      metadataBytes: jcs(metadata),
    });
  } catch (e) {
    return fail(
      report,
      "METADATA_NOT_CANONICALIZABLE",
      `metadata.json cannot be represented as RFC 8785 JCS: ${(e as Error).message}.`,
      metadata,
    );
  }
  const packagedCanonical = entries["canonical.bin"]!;
  if (constantTimeEqual(profileCanonical, packagedCanonical)) {
    report.push("Canonical bytes match AEP profile canonical form.");
  } else if (constantTimeEqual(entries["response.txt"]!, packagedCanonical)) {
    report.push(
      "Canonical bytes match legacy response-only form; metadata is not signature-bound.",
    );
  } else {
    return fail(
      report,
      "CANONICAL_FORM_MISMATCH",
      "canonical.bin does not match a supported canonical form.",
      metadata,
    );
  }
  const canonicalBytes = packagedCanonical;

  // Hash check.
  const expectedHashHex = TEXT_DEC.decode(entries["hash.sha256"]!).trim().toLowerCase();
  const actualHashBytes = await sha256(canonicalBytes);
  const actualHashHex = toHex(actualHashBytes);
  if (actualHashHex !== expectedHashHex) {
    return fail(report, "HASH_MISMATCH", "Hash mismatch.", metadata);
  }
  report.push("SHA-256 hash matches.");

  // RSA signature verification.
  const rsaPem = TEXT_DEC.decode(entries["public_key.pem"]!);
  if (opts.trustedSignerPems && opts.trustedSignerPems.length > 0) {
    const packagedKey = normalizedPemBody(rsaPem);
    const trusted = opts.trustedSignerPems.some(
      (pem) => normalizedPemBody(pem) === packagedKey,
    );
    if (!trusted) {
      return fail(
        report,
        "SIGNER_NOT_TRUSTED",
        "Signer public key is not in the caller-supplied trust set.",
        metadata,
      );
    }
    report.push("Signer public key matched the caller-supplied trust set.");
  } else {
    report.push("Signer identity trust not evaluated (no trustedSignerPems supplied).");
  }
  const rsaSigB64 = TEXT_DEC.decode(entries["signature.sig"]!).trim();
  let rsaSig: Uint8Array;
  try {
    rsaSig = decodeBase64(rsaSigB64);
  } catch (e) {
    return fail(
      report,
      "RSA_SIGNATURE_ENCODING",
      `signature.sig is not valid base64: ${(e as Error).message}.`,
      metadata,
    );
  }
  let rsaOk: boolean;
  let rsaKey: CryptoKey;
  try {
    rsaKey = await importRsaPublicKey(rsaPem);
    rsaOk = await verifyRsa(rsaKey, rsaSig, canonicalBytes);
    if (!rsaOk) {
      rsaOk = verifyRsaDigestInfo(rsaPem, rsaSig, actualHashBytes);
    }
  } catch (e) {
    return fail(
      report,
      "RSA_VERIFY_ERROR",
      `RSA verify error: ${(e as Error).message}.`,
      metadata,
    );
  }
  if (!rsaOk) {
    return fail(
      report,
      "RSA_SIGNATURE_INVALID",
      "RSA signature does not verify against public_key.pem.",
      metadata,
    );
  }
  report.push("RSA-4096 signature verified.");

  const overt = parseAndValidateOvertReceipt(entries, metadata, expectedHashHex);
  overtReceipt = overt.receipt;
  if (overt.error) {
    return fail(
      report,
      "OVERT_INVALID",
      `overt_receipt.json invalid: ${overt.error}.`,
      metadata,
      null,
      overtReceipt,
    );
  }
  report.push(
    overtReceipt
      ? `OVERT receipt verified (${String(overtReceipt.scope)}).`
      : "OVERT receipt absent (optional profile entry).",
  );
  const receiptSignatureName = metadata.overt_receipt_signature;
  if (receiptSignatureName !== undefined) {
    if (receiptSignatureName !== "overt_receipt.sig") {
      return fail(
        report,
        "OVERT_SIGNATURE_MARKER_INVALID",
        "metadata.overt_receipt_signature must equal overt_receipt.sig.",
        metadata,
        null,
        overtReceipt,
      );
    }
    const receiptBytes = entries["overt_receipt.json"];
    const receiptSignatureBytes = entries["overt_receipt.sig"];
    if (!receiptBytes || !receiptSignatureBytes || !overtReceipt) {
      return fail(
        report,
        "OVERT_SIGNATURE_REQUIRED",
        "Signed metadata requires overt_receipt.json and overt_receipt.sig.",
        metadata,
        null,
        overtReceipt,
      );
    }
    try {
      const receiptSig = decodeBase64(TEXT_DEC.decode(receiptSignatureBytes).trim());
      if (!(await verifyRsa(rsaKey, receiptSig, receiptBytes))) {
        return fail(
          report,
          "OVERT_SIGNATURE_INVALID",
          "OVERT receipt signature does not verify against public_key.pem.",
          metadata,
          null,
          overtReceipt,
        );
      }
    } catch (e) {
      return fail(
        report,
        "OVERT_SIGNATURE_ERROR",
        `OVERT receipt signature verify error: ${(e as Error).message}.`,
        metadata,
        null,
        overtReceipt,
      );
    }
    report.push("OVERT receipt signature verified (required by signed metadata).");
  } else if (entries["overt_receipt.sig"]) {
    return fail(
      report,
      "OVERT_SIGNATURE_UNMARKED",
      "Unmarked overt_receipt.sig is not accepted.",
      metadata,
      null,
      overtReceipt,
    );
  } else if (overtReceipt) {
    report.push(
      "Legacy OVERT receipt is cross-checked but not separately signature-bound.",
    );
  }

  // Optional ML-DSA-65 verification.
  let pqcValid: boolean | null = null;
  const hasPqcSignature = Boolean(entries["signature_pqc.sig"]);
  const hasPqcKey = Boolean(entries["pqc_public_key.pem"]);
  if (hasPqcSignature !== hasPqcKey) {
    return fail(
      report,
      "PQC_PAIR_INCOMPLETE",
      "ML-DSA-65 signature and public-key entries must be supplied together.",
      metadata,
      pqcValid,
      overtReceipt,
    );
  }
  if (!hasPqcSignature && opts.pqcPolicy === "required") {
    return fail(
      report,
      "PQC_SIGNATURE_REQUIRED",
      "Verification policy requires a complete ML-DSA-65 signature pair.",
      metadata,
      pqcValid,
      overtReceipt,
    );
  }
  if (hasPqcSignature && hasPqcKey) {
    const pqcSigB64 = TEXT_DEC.decode(entries["signature_pqc.sig"]!).trim();
    const pqcSig = decodeBase64(pqcSigB64);
    const pqcPem = TEXT_DEC.decode(entries["pqc_public_key.pem"]!);
    try {
      pqcValid = await verifyMlDsa65(pqcPem, pqcSig, canonicalBytes);
      report.push(`ML-DSA-65 signature ${pqcValid ? "verified" : "FAILED"}.`);
      if (!pqcValid) {
        return fail(
          report,
          "PQC_SIGNATURE_INVALID",
          "ML-DSA-65 signature does not verify.",
          metadata,
          pqcValid,
          overtReceipt,
        );
      }
    } catch (e) {
      pqcValid = false;
      return fail(
        report,
        "PQC_VERIFY_ERROR",
        `ML-DSA-65 verify error: ${(e as Error).message}.`,
        metadata,
        pqcValid,
        overtReceipt,
      );
    }
  } else {
    report.push("ML-DSA-65 entries absent (transitional v1 package).");
  }

  // v0.1: full RFC 3161 inspection via pkijs — message
  // imprint, SignerInfo signature against embedded cert, genTime,
  // signer DNs.
  const tsaB64 = TEXT_DEC.decode(entries["timestamp.tsr"]!).trim();
  const tsa = await inspectTsa(tsaB64, expectedHashHex, canonicalBytes);
  if (!tsa.tsaPresent) {
    return fail(
      report,
      "TSA_MISSING_OR_INVALID",
      "timestamp.tsr missing or empty.",
      metadata,
      pqcValid,
      overtReceipt,
    );
  }
  report.push(
    `RFC 3161 timestamp present (${tsa.rawSizeBytes} bytes, genTime=${
      tsa.genTime ? tsa.genTime.toISOString() : "unknown"
    }). ` +
      `Message imprint match: ${
        tsa.messageImprintMatches == null ? "could not determine" : tsa.messageImprintMatches
      }. SignerInfo signature: ${
        tsa.signatureVerified == null ? "no embedded cert" : tsa.signatureVerified
      }. Signer: ${tsa.signerSubject ?? "unknown"} (issued by ${tsa.signerIssuer ?? "unknown"}).`,
  );
  if (tsa.messageImprintMatches !== true) {
    return fail(
      report,
      tsa.messageImprintMatches === false
        ? "TSA_IMPRINT_MISMATCH"
        : "TSA_IMPRINT_UNVERIFIABLE",
      tsa.messageImprintMatches === false
        ? "RFC 3161 message imprint does not match hash.sha256."
        : "RFC 3161 message imprint could not be validated as SHA-256.",
      metadata,
      pqcValid,
      overtReceipt,
    );
  }
  if (tsa.signatureVerified !== true) {
    return fail(
      report,
      tsa.signatureVerified === false
        ? "TSA_SIGNATURE_INVALID"
        : "TSA_CERT_MISSING",
      tsa.signatureVerified === false
        ? "RFC 3161 SignerInfo signature did not verify against the embedded certificate."
        : "RFC 3161 token does not contain a verifiable embedded signing certificate.",
      metadata,
      pqcValid,
      overtReceipt,
    );
  }

  // Optional issuer-name pin. This is intentionally not described as
  // RFC 5280 path validation: verifyTsaTrust compares the embedded
  // signer's issuer DN with explicitly supplied root subjects.
  let tsaTrusted: boolean | null = null;
  const trustList = opts.tsaTrustList ?? [];
  if (trustList.length > 0) {
    const trust = await verifyTsaTrust(tsa, trustList);
    tsaTrusted = trust.trusted;
    report.push(
      `TSA issuer-name pin: matched=${trust.trusted}. ${trust.reason}`,
    );
  }

  return {
    valid: true,
    report,
    failureReason: null,
    failureCode: null,
    pqcValid,
    metadata,
    overtReceipt,
    tsaTrusted,
  };
}

function fail(
  report: string[],
  failureCode: FailureCode,
  failureReason: string,
  metadata: Record<string, unknown> | null,
  pqcValid: boolean | null = null,
  overtReceipt: OvertReceipt | null = null,
): VerifyResult {
  report.push("FAIL: " + failureReason);
  return {
    valid: false,
    report,
    failureReason,
    failureCode,
    pqcValid,
    metadata,
    overtReceipt,
  };
}

async function toBytes(input: Uint8Array | ArrayBuffer | Blob): Promise<Uint8Array> {
  if (input instanceof Uint8Array) return input;
  if (input instanceof ArrayBuffer) return new Uint8Array(input);
  if (typeof Blob !== "undefined" && input instanceof Blob) {
    return new Uint8Array(await input.arrayBuffer());
  }
  throw new Error("Unsupported input type for verifier; pass Uint8Array, ArrayBuffer, or Blob.");
}

function constantTimeEqual(a: Uint8Array, b: Uint8Array): boolean {
  if (a.length !== b.length) return false;
  let diff = 0;
  for (let i = 0; i < a.length; i++) {
    diff |= (a[i]! ^ b[i]!);
  }
  return diff === 0;
}

function normalizedPemBody(pem: string): string {
  return pem
    .split(/\r?\n/)
    .filter((line) => !line.startsWith("-----"))
    .join("")
    .replace(/\s+/g, "");
}

File: lib-python/eatf_verifier/verifier.py
"""Top-level verifier — Python port of lib/src/verifier.ts.

Public entrypoint:

    from eatf_verifier import verify
    result = verify(b"...")           # accepts raw bytes
"""

from __future__ import annotations

import base64
import io
import json
import zipfile
from dataclasses import dataclass, field
from typing import Any

from .canonical import jcs
from .hash import sha256, to_hex
from .overt import parse_and_validate_overt_receipt
from .rsa import load_public_key_pem, verify_rsa, verify_rsa_digest_info
from .tsa import inspect_tsa, verify_tsa_trust


@dataclass
class VerifyOptions:
    """Caller-supplied verification configuration."""

    offline_only: bool = True
    trusted_signer_pems: list[bytes] = field(default_factory=list)
    """Optional exact SPKI PEM trust set for the package signer."""
    tsa_trust_list: list[bytes] = field(default_factory=list)
    """Optional PEM certificates for the advisory TSA issuer-name pin.

    An empty list skips the pin. This is not full RFC 5280 path
    validation; see ``verify_tsa_trust``.
    """


@dataclass
class VerifyResult:
    valid: bool
    report: list[str] = field(default_factory=list)
    failure_reason: str | None = None
    pqc_valid: bool | None = None
    tsa_trusted: bool | None = None
    metadata: dict[str, Any] | None = None
    overt_receipt: dict[str, Any] | None = None


REQUIRED_ENTRIES = (
    "response.txt",
    "canonical.bin",
    "hash.sha256",
    "signature.sig",
    "public_key.pem",
    "metadata.json",
    "timestamp.tsr",
)
MAX_ARCHIVE_BYTES = 64 * 1024 * 1024
MAX_ENTRY_BYTES = 16 * 1024 * 1024
MAX_EXPANDED_BYTES = 32 * 1024 * 1024
MAX_ENTRIES = 32


def verify(data: bytes, options: VerifyOptions | None = None) -> VerifyResult:
    """Verify an .aep package. Returns a VerifyResult."""
    opts = options or VerifyOptions()
    report: list[str] = []
    metadata: dict[str, Any] | None = None

    # 1. Unzip with explicit resource and name limits.
    if len(data) > MAX_ARCHIVE_BYTES:
        return _fail(report, "Package exceeds the 64 MiB archive safety limit.", metadata)
    try:
        with zipfile.ZipFile(io.BytesIO(data), "r") as zf:
            infos = zf.infolist()
            names = [info.filename for info in infos]
            if len(infos) > MAX_ENTRIES:
                raise ValueError("too many ZIP entries")
            if len(set(names)) != len(names):
                raise ValueError("duplicate ZIP entry")
            if any(
                "/" in name or "\\" in name or "\0" in name
                for name in names
            ):
                raise ValueError("AEP entries must use flat, safe names")
            if any(info.file_size > MAX_ENTRY_BYTES for info in infos):
                raise ValueError("ZIP entry too large")
            if sum(info.file_size for info in infos) > MAX_EXPANDED_BYTES:
                raise ValueError("expanded ZIP too large")
            entries = {info.filename: zf.read(info) for info in infos}
    except Exception as exc:
        return _fail(
            report,
            f"Package failed ZIP parsing or safety limits: {exc}.",
            metadata,
        )
    report.append(f"Package unzipped ({len(entries)} entries).")

    # 2. Required entries.
    for name in REQUIRED_ENTRIES:
        if name not in entries:
            return _fail(report, f"Missing required entry: {name}.", metadata)

    # 3. Parse metadata.
    try:
        metadata = json.loads(entries["metadata.json"].decode("utf-8"))
        if not isinstance(metadata, dict):
            return _fail(report, "metadata.json is not a JSON object.", None)
    except Exception:
        return _fail(report, "metadata.json is not valid JSON.", None)
    report.append("metadata.json parsed.")

    # 4. Canonical-form check. Response-only is read-only compatibility for
    # early packages and does not authenticate metadata.
    response = entries["response.txt"]
    canonical = entries["canonical.bin"]
    try:
        profile_canonical = response + b"\n" + jcs(metadata)
    except Exception as exc:
        return _fail(
            report,
            f"metadata.json cannot be represented as RFC 8785 JCS: {exc}.",
            metadata,
        )
    if _ct_eq(profile_canonical, canonical):
        report.append("Canonical bytes match AEP profile canonical form.")
    elif _ct_eq(response, canonical):
        report.append(
            "Canonical bytes match legacy response-only form; metadata is not signature-bound."
        )
    else:
        return _fail(
            report,
            "canonical.bin does not match a supported canonical form.",
            metadata,
        )

    # 5. Hash check.
    expected_hash_hex = entries["hash.sha256"].decode("ascii").strip().lower()
    actual_hash = sha256(canonical)
    actual_hash_hex = to_hex(actual_hash)
    if actual_hash_hex != expected_hash_hex:
        return _fail(report, "Hash mismatch.", metadata)
    report.append("SHA-256 hash matches.")

    # 6. RSA signature.
    pem = entries["public_key.pem"]
    if opts.trusted_signer_pems:
        packaged_key = _normalized_pem_body(pem)
        if not any(
            _normalized_pem_body(trusted) == packaged_key
            for trusted in opts.trusted_signer_pems
        ):
            return _fail(
                report,
                "Signer public key is not in the caller-supplied trust set.",
                metadata,
            )
        report.append("Signer public key matched the caller-supplied trust set.")
    else:
        report.append("Signer identity trust not evaluated (no trusted signer keys supplied).")
    sig_b64 = entries["signature.sig"].decode("ascii").strip()
    try:
        sig = base64.b64decode(sig_b64, validate=False)
    except Exception:
        return _fail(report, "signature.sig is not valid base64.", metadata)
    try:
        key = load_public_key_pem(pem)
        rsa_ok = verify_rsa(key, sig, canonical)
        if not rsa_ok:
            # Java-reference compatibility: BouncyCastle DigestInfo
            # encoding without NULL parameters in the SHA-256
            # AlgorithmIdentifier. Strip padding and compare digests.
            rsa_ok = verify_rsa_digest_info(key, sig, actual_hash)
    except Exception as e:
        return _fail(report, f"RSA verify error: {e}.", metadata)
    if not rsa_ok:
        return _fail(
            report, "RSA signature does not verify against public_key.pem.", metadata
        )
    report.append("RSA-4096 signature verified.")

    # 7. OVERT receipt.
    receipt, err = parse_and_validate_overt_receipt(entries, metadata, expected_hash_hex)
    if err:
        return _fail(report, f"overt_receipt.json invalid: {err}.", metadata, None, receipt)
    if receipt is not None:
        report.append(f"OVERT receipt verified ({receipt.get('scope')!s}).")
    else:
        report.append("OVERT receipt absent (optional profile entry).")

    receipt_signature_name = metadata.get("overt_receipt_signature")
    if receipt_signature_name is not None:
        if receipt_signature_name != "overt_receipt.sig":
            return _fail(
                report,
                "metadata.overt_receipt_signature must equal overt_receipt.sig.",
                metadata,
                None,
                receipt,
            )
        receipt_bytes = entries.get("overt_receipt.json")
        receipt_signature_bytes = entries.get("overt_receipt.sig")
        if receipt is None or not receipt_bytes or not receipt_signature_bytes:
            return _fail(
                report,
                "Signed metadata requires overt_receipt.json and overt_receipt.sig.",
                metadata,
                None,
                receipt,
            )
        try:
            receipt_signature = base64.b64decode(
                receipt_signature_bytes.decode("ascii").strip(),
                validate=False,
            )
            if not verify_rsa(key, receipt_signature, receipt_bytes):
                return _fail(
                    report,
                    "OVERT receipt signature does not verify against public_key.pem.",
                    metadata,
                    None,
                    receipt,
                )
        except Exception as exc:
            return _fail(
                report,
                f"OVERT receipt signature verify error: {exc}.",
                metadata,
                None,
                receipt,
            )
        report.append("OVERT receipt signature verified (required by signed metadata).")
    elif "overt_receipt.sig" in entries:
        return _fail(
            report,
            "Unmarked overt_receipt.sig is not accepted.",
            metadata,
            None,
            receipt,
        )
    elif receipt is not None:
        report.append(
            "Legacy OVERT receipt is cross-checked but not separately signature-bound."
        )

    # 8. Optional ML-DSA-65 verification.
    pqc_valid: bool | None = None
    if entries.get("signature_pqc.sig") and entries.get("pqc_public_key.pem"):
        try:
            from .mldsa import verify_mldsa65

            pqc_sig_b64 = entries["signature_pqc.sig"].decode("ascii").strip()
            pqc_sig = base64.b64decode(pqc_sig_b64, validate=False)
            pqc_valid = verify_mldsa65(
                entries["pqc_public_key.pem"], pqc_sig, canonical
            )
            report.append(f"ML-DSA-65 signature {'verified' if pqc_valid else 'FAILED'}.")
            if not pqc_valid:
                return _fail(
                    report,
                    "ML-DSA-65 signature does not verify.",
                    metadata,
                    pqc_valid,
                    receipt,
                )
        except ImportError as e:
            return _fail(
                report,
                f"ML-DSA-65 support not compiled in: {e}",
                metadata,
                None,
                receipt,
            )
    else:
        report.append("ML-DSA-65 entries absent (transitional v1 package).")

    # 9. RFC 3161 timestamp inspection.
    tsr_b64 = entries["timestamp.tsr"].decode("ascii", errors="replace").strip()
    tsa = inspect_tsa(tsr_b64, expected_hash_hex)
    if not tsa.tsa_present:
        return _fail(report, "timestamp.tsr missing or empty.", metadata, pqc_valid, receipt)
    report.append(
        f"RFC 3161 timestamp present ({tsa.raw_size_bytes} bytes, genTime={tsa.gen_time}). "
        f"Message imprint match: {tsa.message_imprint_matches}. "
        f"SignerInfo signature: {tsa.signature_verified}. "
        f"Signer: {tsa.signer_subject} (issued by {tsa.signer_issuer})."
    )
    if tsa.message_imprint_matches is not True:
        reason = (
            "RFC 3161 message imprint does not match hash.sha256."
            if tsa.message_imprint_matches is False
            else "RFC 3161 message imprint could not be validated as SHA-256."
        )
        return _fail(report, reason, metadata, pqc_valid, receipt)
    if tsa.signature_verified is not True:
        reason = (
            "RFC 3161 SignerInfo signature did not verify against the embedded certificate."
            if tsa.signature_verified is False
            else "RFC 3161 token does not contain a verifiable embedded signing certificate."
        )
        return _fail(
            report,
            reason,
            metadata,
            pqc_valid,
            receipt,
        )

    # 10. Optional issuer-name pin (not full RFC 5280 path validation).
    trust_list = opts.tsa_trust_list
    tsa_trusted: bool | None = None
    if trust_list:
        trust = verify_tsa_trust(tsa, trust_list)
        tsa_trusted = trust.trusted
        report.append(f"TSA issuer-name pin: matched={trust.trusted}. {trust.reason}")

    return VerifyResult(
        valid=True,
        report=report,
        failure_reason=None,
        pqc_valid=pqc_valid,
        tsa_trusted=tsa_trusted,
        metadata=metadata,
        overt_receipt=receipt,
    )


def _fail(
    report: list[str],
    reason: str,
    metadata: dict[str, Any] | None,
    pqc_valid: bool | None = None,
    overt_receipt: dict[str, Any] | None = None,
) -> VerifyResult:
    report.append(f"FAIL: {reason}")
    return VerifyResult(
        valid=False,
        report=report,
        failure_reason=reason,
        pqc_valid=pqc_valid,
        tsa_trusted=None,
        metadata=metadata,
        overt_receipt=overt_receipt,
    )


def _ct_eq(a: bytes, b: bytes) -> bool:
    if len(a) != len(b):
        return False
    result = 0
    for x, y in zip(a, b, strict=True):
        result |= x ^ y
    return result == 0


def _normalized_pem_body(pem: bytes) -> bytes:
    return b"".join(
        line.strip()
        for line in pem.splitlines()
        if not line.startswith(b"-----")
    )
